node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -861,17 +930,48 @@ pub(crate) async fn run(
 	let mut runtime = RuntimeInfo::new(None);
 
 	loop {
-		let msg = ctx.recv().fuse().await.map_err(Fatal::SubsystemReceive)?;
-		match msg {
-			Communication { msg } => {
-				log_error(
-					process_msg(&mut ctx, &mut runtime, &mut state, msg).await,
-					"Failed to process message"
-				)?;
+		async fn wait_for_collation_fetch(active: &mut ActiveCollationFetches) -> Hash {
+			loop {
+				if active.is_empty() {
+					futures::pending!()
```

Comment:
I mean you defined that function locally in the same function that uses it with select, but it still looks like standalone code, but it really is not. A comment referring to the select below would help, I think.

Reply:
`pending!()` doesn't automatically wake up. Is that really what you want here?

Reply:
I think it is - look at the `select!` below. There is not much better to do that returning `Pending` in that case and we will be woken up on new messages, which could lead to filling up the `FuturesUnordered` again. I had a similar use in my code just now and ended up writing the future myself with `poll_fn` as it becomes more clear what is going on that way, also then the code is tightly coupled with the other future, which makes their inter-dependency more obvious. If one sticks to those high level primitives, I would not know a better way of writing this either.

Reply:
Yes I want this ;) This `FuturesUnordered` returns `Ready(None)` if it is empty, so not that helpful and can not be filled anywhere in between.

Reply:
You can use `select_next_some` instead fwiw: https://docs.rs/futures/0.3.15/futures/stream/trait.StreamExt.html#method.select_next_some

Reply:
Nice ty! Much easier.